File: 프로그래머스/Lv1/[1차] 비밀지도.py
# 내 풀이
def solution(n, arr1, arr2):
    answer = [bin(i|j) for i, j in zip(arr1, arr2)]
    return [i.replace('0b', '').rjust(n, '0').replace('1', '#').replace('0', ' ') for i in answer]

# 다른 사람 풀이
solution2 = lambda n, arr1, arr2: ([''.join(map(lambda x: '#' if x=='1' else ' ', "{0:b}".format(row).zfill(n))) for row in (a|b for a, b in zip(arr1, arr2))])

print(solution(5, [9, 20, 28, 18, 11], [30, 1, 21, 17, 28]))
print(solution2(6, [46, 33, 33 ,22, 31, 50], [27 ,56, 19, 14, 14, 10]))

File: 프로그래머스/Lv1/[1차] 비밀지도.js
// 내 풀이
function solution(n, arr1, arr2) {
    let answer = [];

    let map1 = secret_map(n, arr1);
    let map2 = secret_map(n, arr2);

    for(let i=0; i<n; i++){
        let temp = [];
        for(let j=0; j<n; j++){
            if(map1[i][j] + map2[i][j] >= 1){
                temp.push('#');
            }else{
                temp.push(' ');
            }
        }
        answer.push(temp.join(''));
    }

    return answer;
}

function secret_map(n, arr){
    let answer = [];

    for(let i=0; i<n; i++){
        let num = arr[i];
        let temp = [];
        for(j=n-1; j>=0; j--){
            if(num >= 2**j){
                num -= 2**j;
                temp.push(1);
            }else{
                temp.push(0);
            }
        }
        answer.push(temp);
    }

    return answer;
}

// 다른 사람 풀이 1
function solution1(n, arr1, arr2) {
    return arr1.map((v, i) => addZero(n, (v | arr2[i]).toString(2)).replace(/1|0/g, a => +a ? '#' : ' '));
}

const addZero = (n, s) => {
    return '0'.repeat(n - s.length) + s;
}

// 다른 사람 풀이 2
var solution2=(n,a,b)=>a.map((a,i)=>(a|b[i]).toString(2).padStart(n,0).replace(/0/g,' ').replace(/1/g,'#'))

console.log(solution(5, [9, 20, 28, 18, 11], [30, 1, 21, 17, 28]));
console.log(solution1(5, [9, 20, 28, 18, 11], [30, 1, 21, 17, 28]));
console.log(solution2(5, [9, 20, 28, 18, 11], [30, 1, 21, 17, 28]));